### ISSUE TITLE
Is there a migration guide for 2.x to 3.x?

### ISSUE THREAD

**Issue by mdhornet90** [okta/okta-idx-swift#121]
### Describe the feature request?

I'm struggling to fit 3.x in the old 2.x paradigm in the code, is there a migration guide available somewhere?

### New or Affected Resource(s)

N/A

**Comment by mikenachbaur-okta** [comment]
We don't have a migration guide per-se, but the biggest changes are:

1. OktaIdx now builds upon [AuthFoundation](https://okta.github.io/okta-mobile-swift/development/authfoundation/), which is a part of [okta-mobile-swift](https://github.com/okta/okta-mobile-swift/).  This supports built-in token storage and lifecycle management, which wasn't previously available in the IDX library.  For more information about this SDK and its benefits, please refer to an [article I wrote on the Okta Developer Blog.](https://developer.okta.com/blog/2022/08/30/introducing-the-new-okta-mobile-sdks)

2. To fit within the naming convention of the authentication flows provided by the new Mobile SDK, and to better match the naming of the "Interaction Code Flow" grant type used within the Okta Dashboard, the `IDXClient` class was renamed `InteractionCodeFlow`.
![Screen Shot 2022-12-23 at 10 43 31 AM](https://user-images.githubusercontent.com/74688448/209393358-891f554d-1a36-42bb-8247-199bc7222dbc.png)

3. The client object is no longer created asynchronously. Instead you create an instance of [InteractionCodeFlow](https://okta.github.io/okta-idx-swift/development/oktaidx/documentation/oktaidx/interactioncodeflow), and run the [start](https://okta.github.io/okta-idx-swift/development/oktaidx/documentation/oktaidx/interactioncodeflow/start(options:)) function to get your first response.

From that point forward, the usage patterns are the same. You still work with responses and remediations.

If you have any additional questions, please let me know. I'll try to add some improved API documentation to simplify migrations to the 3.x versions.

**Comment by mikenachbaur-okta** [comment]
@mdhornet90 I wanted to check in after the holidays to see if this helped. If this doesn't answer all your questions, would you mind giving me some more information about the problems you're facing, and I'll try to help.

**Comment by mdhornet90** [comment]
Hi @mikenachbaur-okta, I had to put the migration work down as other priorities came up but I'll be addressing this soon. Basically what I was finding was after resolving compiler errors, basic flows I had coded up against 3.X.X were returning errors I hadn't seen before, specifically the cryptic `the data cannot be read because it is missing`

I'm sure these problems are highly specific to a given app's setup, so I can post some sanitized code if that makes troubleshooting easier, just lmk.

**Comment by mikenachbaur-okta** [comment]
Interesting @mdhornet90, that error is reported when JSON decoding is unable to find a certain key. Would you be able to share the details of that error you're getting? Perhaps, if you can capture that in the debugger, where in the call stack it's being reported?

**Comment by mdhornet90** [comment]
Here's the code up to where the request fails (the last line), note that the code itself was minimally modified just to resolve compiler errors as a result of migrating from 2.x->3.x:
```
    public func signIn(userName: String, password: String) async throws -> SignInResult {
        try await withErrorHandler {
            let client = OktaAuthenticationService.createClient() // Creates client with necessary arguments
            let resumeResponse = try await client.start()
            guard
                let identifyRemediation = resumeResponse.remediations[.identify],
                let usernameField = identifyRemediation["identifier"]
            else {
                throw AuthenticationError.unexpectedError("Unexpected error occurred when identifying user during sign in")
            }
            usernameField.value = userName
            
            let identifyResponse = try await identifyRemediation.proceed()
            if identifyResponse.remediations[.selectAuthenticatorUnlockAccount] != nil {
                throw AuthenticationError.unexpectedError("Your account has been locked, please attempt account recovery")
            }
            guard
                let challengeAuthenticatorRemediation = identifyResponse.remediations[.challengeAuthenticator],
                let passwordField = challengeAuthenticatorRemediation["credentials.passcode"]
            else {
                throw AuthenticationError.unexpectedError("Unexpected error occured validating password during sign in")
            }
            passwordField.value = password
            
            let challengeResponse = try await challengeAuthenticatorRemediation.proceed() // FAILS HERE
 ```
 Here's a screenshot of the stack trace at the time of the failure: 
<img width="572" alt="image" src="https://user-images.githubusercontent.com/424092/212777420-e9d2018a-d976-49a6-a272-2b2096fad616.png">

And here's a screenshot of the code in `AuthFoundation` it's specifically failing on:
<img width="748" alt="image" src="https://user-images.githubusercontent.com/424092/212777485-327e5884-91ad-429a-a59c-6670f7eaf6ca.png">

Let me know if you need anything else to assist with debugging this, nothing is spat out in the console from what I can tell.

**Comment by mikenachbaur-okta** [comment]
I'll try implementing a test locally that follows a similar pattern to what you're doing, and I'll see if I can reproduce your problem on my system. However, individual policy settings within an organization has a big impact on the data returned, so I'm not certain if I'll be able to hit the same issue as you are.

If you're able to sanitize the JSON data (so no interaction handles or PII is included), would you be able to share this with me? You can set a breakpoint in the `APIClient/send<T>(_:parsing:state:completion:)` function to inspect the returned data (perhaps around [this line here](https://github.com/okta/okta-mobile-swift/blob/master/Sources/AuthFoundation/Network/APIClient.swift#L193)). From there you can print the contents of `data` to see the raw JSON response. I'd be curious what the JSON result is that is causing this error.

![Screen Shot 2023-01-17 at 11 28 28 AM](https://user-images.githubusercontent.com/74688448/212994007-0fe60dbd-1c20-4811-904a-5ca7900397de.png)

**Comment by mdhornet90** [comment]
That was super helpful, thanks for the suggestion. Here's the json data I see:
```
{
    "version": "1.0.0",
    "stateHandle": "02.id.uNYyG7V3juiCezXNCSarHQCyOXWjtpQFgLT8givM~c.ugHG791iSmbMrGjr831VuXVwABPzNMXRf9R2I9WNlPYmyxjMWjosYEC2uB6CILB0TSPuud56_GfiYWmQD0BZc1v3AZiX2IULGXdyo3TLPsY44bY1TrxRfri1NBvYtMVlxmnva6ztsc_3epy88x7myhSvNh11TYyABphgdGgRNY7VV9pnzLmr_lASFjxdZp-tHzi5DB3BORM359YERWocCv8ihoXRtZUM-aHIMLletux0aQLpgrRn61ruX06g-sQr8o-CG7GUtOUFAeGsurkXPdvzeZc3omI7uaCs3Dt9jHTjyWiKSFyzL4n54qDXhnCQlt0fZFuzf73FJL3nVU_hjCMpOAAx0i4as2RHE_g1Hmbll_FNfV-TKk6hV5jLUMGlUfwdV-bd3SXLp8g-B4KvRwwz87t2cuAkjeHbqcDH7_JdLTh56aQKa7ofv9KbwSFkZaNS8udG6P19pQ_iDsOyRPUwSp_i05orEJnXpf_Jo9i36nkgHp_t2p1Szb-niWuGBK0nBA8Sh5XCfy7Wf6xTu2W_MeEpq3_dfJIiVTIy1ga3FXlEu61bg6H2r7ACHyFl28RaQ2144O-NKrMGc33fPwzI7ArbIZ5bxjwm3F4UadMiZPC9dpZOeaaLfnVBZClp5Efw3g_V-nn3yj5AKMR_BpacQ5Myt1vKX0VCzlEKLFOHzYNh4VuJPUrPXimnCLKSUSjZM9AAPBhoykW4IBDu14ApdG0VADYNOMGBpbCcDUGXYL2JhLUFzH-hpPA_vwSNbY-vLTGdXGhJUluYJquw6xUyFuaYhxo4oQ99ryOBGBGp77_HHArYGfJQRbE7b6KV",
    "expiresAt": "2023-01-18T21:37:39.000Z",
    "intent": "LOGIN",
    "remediation": {
        "type": "array",
        "value": [
            {
                "rel": [
                    "create-form"
                ],
                "name": "authenticator-verification-data",
                "relatesTo": [
                    "$.currentAuthenticatorEnrollment"
                ],
                "href": "https://sandbox-inc-eng.oktapreview.com/idp/idx/challenge",
                "method": "POST",
                "produces": "application/ion+json; okta-version=1.0.0",
                "value": [
                    {
                        "name": "authenticator",
                        "label": "Phone",
                        "form": {
                            "value": [
                                {
                                    "name": "id",
                                    "required": true,
                                    "value": "aut3et3uzjUoJxnB31d7",
                                    "mutable": false
                                },
                                {
                                    "name": "methodType",
                                    "type": "string",
                                    "required": true,
                                    "options": [
                                        {
                                            "label": "SMS",
                                            "value": "sms"
                                        }
                                    ]
                                },
                                {
                                    "name": "enrollmentId",
                                    "required": true,
                                    "value": "sms3r68zjsGBZp30t1d7",
                                    "mutable": false
                                }
                            ]
                        }
                    },
                    {
                        "name": "stateHandle",
                        "required": true,
                        "value": "02.id.uNYyG7V3juiCezXNCSarHQCyOXWjtpQFgLT8givM~c.ugHG791iSmbMrGjr831VuXVwABPzNMXRf9R2I9WNlPYmyxjMWjosYEC2uB6CILB0TSPuud56_GfiYWmQD0BZc1v3AZiX2IULGXdyo3TLPsY44bY1TrxRfri1NBvYtMVlxmnva6ztsc_3epy88x7myhSvNh11TYyABphgdGgRNY7VV9pnzLmr_lASFjxdZp-tHzi5DB3BORM359YERWocCv8ihoXRtZUM-aHIMLletux0aQLpgrRn61ruX06g-sQr8o-CG7GUtOUFAeGsurkXPdvzeZc3omI7uaCs3Dt9jHTjyWiKSFyzL4n54qDXhnCQlt0fZFuzf73FJL3nVU_hjCMpOAAx0i4as2RHE_g1Hmbll_FNfV-TKk6hV5jLUMGlUfwdV-bd3SXLp8g-B4KvRwwz87t2cuAkjeHbqcDH7_JdLTh56aQKa7ofv9KbwSFkZaNS8udG6P19pQ_iDsOyRPUwSp_i05orEJnXpf_Jo9i36nkgHp_t2p1Szb-niWuGBK0nBA8Sh5XCfy7Wf6xTu2W_MeEpq3_dfJIiVTIy1ga3FXlEu61bg6H2r7ACHyFl28RaQ2144O-NKrMGc33fPwzI7ArbIZ5bxjwm3F4UadMiZPC9dpZOeaaLfnVBZClp5Efw3g_V-nn3yj5AKMR_BpacQ5Myt1vKX0VCzlEKLFOHzYNh4VuJPUrPXimnCLKSUSjZM9AAPBhoykW4IBDu14ApdG0VADYNOMGBpbCcDUGXYL2JhLUFzH-hpPA_vwSNbY-vLTGdXGhJUluYJquw6xUyFuaYhxo4oQ99ryOBGBGp77_HHArYGfJQRbE7b6KV",
                        "visible": false,
                        "mutable": false
                    }
                ],
                "accepts": "application/json; okta-version=1.0.0"
            },
            {
                "rel": [
                    "create-form"
                ],
                "name": "select-authenticator-authenticate",
                "href": "https://sandbox-inc-eng.oktapreview.com/idp/idx/challenge",
                "method": "POST",
                "produces": "application/ion+json; okta-version=1.0.0",
                "value": [
                    {
                        "name": "authenticator",
                        "type": "object",
                        "options": [
                            {
                                "label": "Phone",
                                "value": {
                                    "form": {
                                        "value": [
                                            {
                                                "name": "id",
                                                "required": true,
                                                "value": "aut3et3uzjUoJxnB31d7",
                                                "mutable": false
                                            },
                                            {
                                                "name": "methodType",
                                                "type": "string",
                                                "required": false,
                                                "options": [
                                                    {
                                                        "label": "SMS",
                                                        "value": "sms"
                                                    }
                                                ]
                                            },
                                            {
                                                "name": "enrollmentId",
                                                "required": true,
                                                "value": "sms3r68zjsGBZp30t1d7",
                                                "mutable": false
                                            }
                                        ]
                                    }
                                },
                                "relatesTo": "$.authenticatorEnrollments.value[0]"
                            }
                        ]
                    },
                    {
                        "name": "stateHandle",
                        "required": true,
                        "value": "02.id.uNYyG7V3juiCezXNCSarHQCyOXWjtpQFgLT8givM~c.ugHG791iSmbMrGjr831VuXVwABPzNMXRf9R2I9WNlPYmyxjMWjosYEC2uB6CILB0TSPuud56_GfiYWmQD0BZc1v3AZiX2IULGXdyo3TLPsY44bY1TrxRfri1NBvYtMVlxmnva6ztsc_3epy88x7myhSvNh11TYyABphgdGgRNY7VV9pnzLmr_lASFjxdZp-tHzi5DB3BORM359YERWocCv8ihoXRtZUM-aHIMLletux0aQLpgrRn61ruX06g-sQr8o-CG7GUtOUFAeGsurkXPdvzeZc3omI7uaCs3Dt9jHTjyWiKSFyzL4n54qDXhnCQlt0fZFuzf73FJL3nVU_hjCMpOAAx0i4as2RHE_g1Hmbll_FNfV-TKk6hV5jLUMGlUfwdV-bd3SXLp8g-B4KvRwwz87t2cuAkjeHbqcDH7_JdLTh56aQKa7ofv9KbwSFkZaNS8udG6P19pQ_iDsOyRPUwSp_i05orEJnXpf_Jo9i36nkgHp_t2p1Szb-niWuGBK0nBA8Sh5XCfy7Wf6xTu2W_MeEpq3_dfJIiVTIy1ga3FXlEu61bg6H2r7ACHyFl28RaQ2144O-NKrMGc33fPwzI7ArbIZ5bxjwm3F4UadMiZPC9dpZOeaaLfnVBZClp5Efw3g_V-nn3yj5AKMR_BpacQ5Myt1vKX0VCzlEKLFOHzYNh4VuJPUrPXimnCLKSUSjZM9AAPBhoykW4IBDu14ApdG0VADYNOMGBpbCcDUGXYL2JhLUFzH-hpPA_vwSNbY-vLTGdXGhJUluYJquw6xUyFuaYhxo4oQ99ryOBGBGp77_HHArYGfJQRbE7b6KV",
                        "visible": false,
                        "mutable": false
                    }
                ],
                "accepts": "application/json; okta-version=1.0.0"
            }
        ]
    },
    "currentAuthenticatorEnrollment": {
        "type": "object",
        "value": {
            "profile": {
                "phoneNumber": "+1 XXX-XXX-0990"
            },
            "resend": {
                "rel": [
                    "create-form"
                ],
                "name": "resend",
                "href": "https://sandbox-inc-eng.oktapreview.com/idp/idx/challenge/resend",
                "method": "POST",
                "produces": "application/ion+json; okta-version=1.0.0",
                "value": [
                    {
                        "name": "stateHandle",
                        "required": true,
                        "value": "02.id.uNYyG7V3juiCezXNCSarHQCyOXWjtpQFgLT8givM~c.ugHG791iSmbMrGjr831VuXVwABPzNMXRf9R2I9WNlPYmyxjMWjosYEC2uB6CILB0TSPuud56_GfiYWmQD0BZc1v3AZiX2IULGXdyo3TLPsY44bY1TrxRfri1NBvYtMVlxmnva6ztsc_3epy88x7myhSvNh11TYyABphgdGgRNY7VV9pnzLmr_lASFjxdZp-tHzi5DB3BORM359YERWocCv8ihoXRtZUM-aHIMLletux0aQLpgrRn61ruX06g-sQr8o-CG7GUtOUFAeGsurkXPdvzeZc3omI7uaCs3Dt9jHTjyWiKSFyzL4n54qDXhnCQlt0fZFuzf73FJL3nVU_hjCMpOAAx0i4as2RHE_g1Hmbll_FNfV-TKk6hV5jLUMGlUfwdV-bd3SXLp8g-B4KvRwwz87t2cuAkjeHbqcDH7_JdLTh56aQKa7ofv9KbwSFkZaNS8udG6P19pQ_iDsOyRPUwSp_i05orEJnXpf_Jo9i36nkgHp_t2p1Szb-niWuGBK0nBA8Sh5XCfy7Wf6xTu2W_MeEpq3_dfJIiVTIy1ga3FXlEu61bg6H2r7ACHyFl28RaQ2144O-NKrMGc33fPwzI7ArbIZ5bxjwm3F4UadMiZPC9dpZOeaaLfnVBZClp5Efw3g_V-nn3yj5AKMR_BpacQ5Myt1vKX0VCzlEKLFOHzYNh4VuJPUrPXimnCLKSUSjZM9AAPBhoykW4IBDu14ApdG0VADYNOMGBpbCcDUGXYL2JhLUFzH-hpPA_vwSNbY-vLTGdXGhJUluYJquw6xUyFuaYhxo4oQ99ryOBGBGp77_HHArYGfJQRbE7b6KV",
                        "visible": false,
                        "mutable": false
                    }
                ],
                "accepts": "application/json; okta-version=1.0.0"
            },
            "type": "phone",
            "key": "phone_number",
            "id": "sms3r68zjsGBZp30t1d7",
            "displayName": "Phone",
            "methods": [
                {
                    "type": "sms"
                }
            ]
        }
    },
    "authenticators": {
        "type": "array",
        "value": [
            {
                "type": "phone",
                "key": "phone_number",
                "id": "aut3et3uzjUoJxnB31d7",
                "displayName": "Phone",
                "methods": [
                    {
                        "type": "sms"
                    }
                ],
                "allowedFor": "any"
            }
        ]
    },
    "authenticatorEnrollments": {
        "type": "array",
        "value": [
            {
                "profile": {
                    "phoneNumber": "+1 XXX-XXX-0990"
                },
                "type": "phone",
                "key": "phone_number",
                "id": "sms3r68zjsGBZp30t1d7",
                "displayName": "Phone",
                "methods": [
                    {
                        "type": "sms"
                    }
                ]
            }
        ]
    },
    "user": {
        "type": "object",
        "value": {
            "id": "00u3r66s0wzUtzAdm1d7",
            "identifier": "my.email@foo.com",
            "profile": {
                "firstName": null,
                "lastName": null,
                "timeZone": "America/Los_Angeles",
                "locale": "en_US"
            }
        }
    },
    "cancel": {
        "rel": [
            "create-form"
        ],
        "name": "cancel",
        "href": "https://sandbox-inc-eng.oktapreview.com/idp/idx/cancel",
        "method": "POST",
        "produces": "application/ion+json; okta-version=1.0.0",
        "value": [
            {
                "name": "stateHandle",
                "required": true,
                "value": "02.id.uNYyG7V3juiCezXNCSarHQCyOXWjtpQFgLT8givM~c.ugHG791iSmbMrGjr831VuXVwABPzNMXRf9R2I9WNlPYmyxjMWjosYEC2uB6CILB0TSPuud56_GfiYWmQD0BZc1v3AZiX2IULGXdyo3TLPsY44bY1TrxRfri1NBvYtMVlxmnva6ztsc_3epy88x7myhSvNh11TYyABphgdGgRNY7VV9pnzLmr_lASFjxdZp-tHzi5DB3BORM359YERWocCv8ihoXRtZUM-aHIMLletux0aQLpgrRn61ruX06g-sQr8o-CG7GUtOUFAeGsurkXPdvzeZc3omI7uaCs3Dt9jHTjyWiKSFyzL4n54qDXhnCQlt0fZFuzf73FJL3nVU_hjCMpOAAx0i4as2RHE_g1Hmbll_FNfV-TKk6hV5jLUMGlUfwdV-bd3SXLp8g-B4KvRwwz87t2cuAkjeHbqcDH7_JdLTh56aQKa7ofv9KbwSFkZaNS8udG6P19pQ_iDsOyRPUwSp_i05orEJnXpf_Jo9i36nkgHp_t2p1Szb-niWuGBK0nBA8Sh5XCfy7Wf6xTu2W_MeEpq3_dfJIiVTIy1ga3FXlEu61bg6H2r7ACHyFl28RaQ2144O-NKrMGc33fPwzI7ArbIZ5bxjwm3F4UadMiZPC9dpZOeaaLfnVBZClp5Efw3g_V-nn3yj5AKMR_BpacQ5Myt1vKX0VCzlEKLFOHzYNh4VuJPUrPXimnCLKSUSjZM9AAPBhoykW4IBDu14ApdG0VADYNOMGBpbCcDUGXYL2JhLUFzH-hpPA_vwSNbY-vLTGdXGhJUluYJquw6xUyFuaYhxo4oQ99ryOBGBGp77_HHArYGfJQRbE7b6KV",
                "visible": false,
                "mutable": false
            }
        ],
        "accepts": "application/json; okta-version=1.0.0"
    },
    "app": {
        "type": "object",
        "value": {
            "name": "oidc_client",
            "label": "IOS App",
            "id": "0oa3mbs714hH4yNvd1d7"
        }
    },
    "authentication": {
        "type": "object",
        "value": {
            "protocol": "OAUTH2.0",
            "issuer": {
                "id": "aus3ew5mocPtYxZmf1d7",
                "name": "default",
                "uri": "https://sandbox-inc-eng.oktapreview.com/oauth2/default"
            },
            "request": {
                "max_age": -1,
                "scope": "openid profile offline_access",
                "response_type": "code",
                "redirect_uri": "com.oktapreview.sandbox-inc-eng:/callback",
                "state": "7E5E700F-2E39-451D-9F61-E18E394AB4BE",
                "code_challenge_method": "S256",
                "code_challenge": "I6bXAkujVNASoIzcEJD_U-MHGsf_NwVlOUvImmKRceE",
                "response_mode": "query"
            }
        }
    }
}
```

Does this help at all?

**Comment by mdhornet90** [comment]
I traced a little further and found this error:
```
▿ DecodingError
  ▿ valueNotFound : 2 elements
    - .0 : Swift.String
    ▿ .1 : Context
      ▿ codingPath : 4 elements
        - 0 : CodingKeys(stringValue: "user", intValue: nil)
        - 1 : CodingKeys(stringValue: "value", intValue: nil)
        - 2 : CodingKeys(stringValue: "profile", intValue: nil)
        ▿ 3 : _JSONKey(stringValue: "firstName", intValue: nil)
          - stringValue : "firstName"
          - intValue : nil
      - debugDescription : "Expected String but found null value instead."
      - underlyingError : nil
      ```
And sure enough the test user I'm attempting to log in with has a blank first name and last name. Any idea why this would start mattering in the latest version of the SDK? For reference the scopes I'm requesting are `openid profile offline_access`

**Comment by mikenachbaur-okta** [comment]
@mdhornet90 That is _exactly_ the information I needed, thank you. The [User Profile properties are marked as optional](https://github.com/okta/okta-idx-swift/blob/master/Sources/OktaIdx/IDXUser.swift#L35), but it seems like the JSONDecoder is having an issue with the key being included in the JSON with a null value.

I've incorporated this into a test case (which is currently failing locally), and will get a fix out to you as soon as I can.

**Comment by mikenachbaur-okta** [comment]
@mdhornet90 I just [released OktaIdx 3.0.7](https://github.com/okta/okta-idx-swift/releases/tag/3.0.7) which should address this issue. If it's still causing problems, please feel free to reopen this ticket, and I'd be glad to help track the problem down.

**Comment by mdhornet90** [comment]
Finally had a chance to look at this and 3.0.7 indeed resolves the issue! There's still more work to be done on repairing token management due to the AuthFoundation changes, but thanks for your help with this @mikenachbaur-okta
